lib/semantics/check-coarray.h

```diff
@@ -41,5 +44,40 @@ class CoarrayChecker : public virtual BaseChecker {
       const parser::CharBlock &, parser::MessageFixedText &&);
 };
 
+class CriticalBodyEnforce {
+public:
+  CriticalBodyEnforce(
+      SemanticsContext &context, parser::CharBlock criticalSourcePosition)
+    : context_{context}, criticalSourcePosition_{criticalSourcePosition} {}
+  std::set<parser::Label> labels() { return labels_; }
+  template<typename T> bool Pre(const T &) { return true; }
+  template<typename T> void Post(const T &) {}
+
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    currentStatementSourcePosition_ = statement.source;
+    if (statement.label.has_value()) {
+      labels_.insert(*statement.label);
+    }
+    return true;
+  }
+
+  // C1118
+  void Post(const parser::ReturnStmt &) {
+    context_.Say(currentStatementSourcePosition_,
+        "RETURN statement is not allowed in a CRITICAL construct"_err_en_US);
+  }
+  void Post(const parser::ExecutableConstruct &construct) {
+    if (IsImageControlStmt(construct)) {
+      context_.Say(currentStatementSourcePosition_,
+          "An image control statement is not allowed in a CRITICAL construct"_err_en_US);
```

Comment:
Can you split this so that the lines fit into 80 characters?  Unfortunately, clang-format doesn't do this automatically.